kafka/handle.go

```diff
@@ -247,26 +309,44 @@ type cgoDr struct {
 
 // cgoPut adds object cg to the handle's cgo map and returns a
 // unique id for the added entry.
+//
+// All void* parameters passed into cgo need to either point at a valid Go
+// object, or something that definitely can't be a valid Go object (i.e. a
+// valid pointer to something in the C heap). Furthermore, an unsafe.Pointer
+// pointing to a Go object cannot be stored in a C structure that is passed
+// to a cgo call.
+//
+// This method therefore returns an unsafe.Pointer that is guaranteed to never
+// point to a valid Go object, by invoking C.malloc to produce it. This can
+// safely be passed through cgo calls and back into cgoGet() to return the
+// passed-in object.
+//
 // Thread-safe.
-// FIXME: the uniquity of the id is questionable over time.
-func (h *handle) cgoPut(cg cgoif) (cgoid int) {
+func (h *handle) cgoPut(cg cgoif) unsafe.Pointer {
```

Comment:
This signature change (of `int` to `unsafe.Pointer`) is needed to avoid repeated cgo calls in the batch produce case as you asked for. I explained this in the commit message but I'll paste it here for easy review too.

    To round-trip values through the _private void* field in the
    rd_kafka_message_t type, we currently generate an int in cgoPut(), save
    the value in a map, and cast the int to a uintptr_t when passing into
    cgo, which is then cast (inside C code) to a void* before sending to
    librdkafka.
    
    To add support for passing _private in the batch-produce case, we need
    to be able to assign the _private field (of type void*/unsafe.Pointer)
    in Go, rather than in C. In Go, it is illegal to cast any int to an
    unsafe.Pointer (even if it is never dereferenced).
    
    Therefore, we must change cgoPut/cgoGet to deal with unsafe.Pointer
    tokens, rather than int tokens. The most straightforward way to generate
    such pointers which are guaranteed to never point to Go objects is to
    allocate them on the C heap with C.malloc.
    
    To avoid a bunch of pointless malloc/free calls, we also implement a
    token cache so that these pointers can be reused.